File: C++/Constraints/bounds.cpp
// Copyright 2019 Alexander Liniger

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
///////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////

#include "bounds.h"
namespace mpcc{
Bounds::Bounds()
{
    std::cout << "default constructor, not everything is initialized properly" << std::endl;
}

Bounds::Bounds(BoundsParam bounds_param) 
{
    l_bounds_x_(0) = bounds_param.lower_state_bounds.X_l;
    l_bounds_x_(1) = bounds_param.lower_state_bounds.Y_l;
    l_bounds_x_(2) = bounds_param.lower_state_bounds.phi_l;
    l_bounds_x_(3) = bounds_param.lower_state_bounds.vx_l;
    l_bounds_x_(4) = bounds_param.lower_state_bounds.vy_l;
    l_bounds_x_(5) = bounds_param.lower_state_bounds.r_l;
    l_bounds_x_(6) = bounds_param.lower_state_bounds.s_l;
    l_bounds_x_(7) = bounds_param.lower_state_bounds.D_l;
    l_bounds_x_(8) = bounds_param.lower_state_bounds.delta_l;
    l_bounds_x_(9) = bounds_param.lower_state_bounds.vs_l;

    u_bounds_x_(0) = bounds_param.upper_state_bounds.X_u;
    u_bounds_x_(1) = bounds_param.upper_state_bounds.Y_u;
    u_bounds_x_(2) = bounds_param.upper_state_bounds.phi_u;
    u_bounds_x_(3) = bounds_param.upper_state_bounds.vx_u;
    u_bounds_x_(4) = bounds_param.upper_state_bounds.vy_u;
    u_bounds_x_(5) = bounds_param.upper_state_bounds.r_u;
    u_bounds_x_(6) = bounds_param.upper_state_bounds.s_u;
    u_bounds_x_(7) = bounds_param.upper_state_bounds.D_u;
    u_bounds_x_(8) = bounds_param.upper_state_bounds.delta_u;
    u_bounds_x_(9) = bounds_param.upper_state_bounds.vs_u;

    l_bounds_u_(0) = bounds_param.lower_input_bounds.dD_l;
    l_bounds_u_(1) = bounds_param.lower_input_bounds.dDelta_l;
    l_bounds_u_(2) = bounds_param.lower_input_bounds.dVs_l;

    u_bounds_u_(0) = bounds_param.upper_input_bounds.dD_u;
    u_bounds_u_(1) = bounds_param.upper_input_bounds.dDelta_u;
    u_bounds_u_(2) = bounds_param.upper_input_bounds.dVs_u;

    l_bounds_s_ = Bounds_s::Zero();
    u_bounds_s_ = Bounds_s::Zero();

    std::cout << "bounds initialized" << std::endl;
}

Bounds_x Bounds::getBoundsLX() const
{
    return  l_bounds_x_;
}

Bounds_x Bounds::getBoundsUX() const
{
    return  u_bounds_x_;
}

Bounds_u Bounds::getBoundsLU() const
{
    return  l_bounds_u_;
}

Bounds_u Bounds::getBoundsUU() const
{
    return  u_bounds_u_;
}

Bounds_s Bounds::getBoundsLS() const
{
    return  l_bounds_s_;
}

Bounds_s Bounds::getBoundsUS() const{
    return  u_bounds_s_;
}
}

File: Matlab/constraints/Bounds.m
classdef Bounds
    %BOUNDS Summary of this class goes here
    %   Detailed explanation goes here
    
    properties (Access = private)
        d_uBoundsX
        d_lBoundsX
        d_uBoundsU
        d_lBoundsU
        d_uBoundsS
        d_lBoundsS
    end
    
    methods (Access = public)
        function obj = Bounds(bounds)
            obj.d_uBoundsX = zeros(11,1);
            obj.d_lBoundsX = zeros(11,1);
            obj.d_uBoundsU = zeros(4,1);
            obj.d_lBoundsU = zeros(4,1);
            obj.d_uBoundsS = zeros(2,1);
            obj.d_lBoundsS = zeros(2,1);

            obj.d_lBoundsX(1) = bounds.lowerStateBounds.xL;
            obj.d_lBoundsX(2) = bounds.lowerStateBounds.yL;
            obj.d_lBoundsX(3) = bounds.lowerStateBounds.phiL;
            obj.d_lBoundsX(4) = bounds.lowerStateBounds.vxL;
            obj.d_lBoundsX(5) = bounds.lowerStateBounds.vyL;
            obj.d_lBoundsX(6) = bounds.lowerStateBounds.rL;
            obj.d_lBoundsX(7) = bounds.lowerStateBounds.sL;
            obj.d_lBoundsX(8) = bounds.lowerStateBounds.throttleL;
            obj.d_lBoundsX(9) = bounds.lowerStateBounds.steeringAngleL;
            obj.d_lBoundsX(10) = bounds.lowerStateBounds.brakesL;
            obj.d_lBoundsX(11) = bounds.lowerStateBounds.vsL;

            obj.d_uBoundsX(1) = bounds.upperStateBounds.xU;
            obj.d_uBoundsX(2) = bounds.upperStateBounds.yU;
            obj.d_uBoundsX(3) = bounds.upperStateBounds.phiU;
            obj.d_uBoundsX(4) = bounds.upperStateBounds.vxU;
            obj.d_uBoundsX(5) = bounds.upperStateBounds.vyU;
            obj.d_uBoundsX(6) = bounds.upperStateBounds.rU;
            obj.d_uBoundsX(7) = bounds.upperStateBounds.sU;
            obj.d_uBoundsX(8) = bounds.upperStateBounds.throttleU;
            obj.d_uBoundsX(9) = bounds.upperStateBounds.steeringAngleU;
            obj.d_uBoundsX(10) = bounds.upperStateBounds.brakesU;
            obj.d_uBoundsX(11) = bounds.upperStateBounds.vsU;

            obj.d_lBoundsU(1) = bounds.lowerInputBounds.dThrottleL;
            obj.d_lBoundsU(2) = bounds.lowerInputBounds.dSteeringAngleL;
            obj.d_lBoundsU(3) = bounds.lowerInputBounds.dBrakesL;
            obj.d_lBoundsU(4) = bounds.lowerInputBounds.dVsL;

            obj.d_uBoundsU(1) = bounds.upperInputBounds.dThrottleU;
            obj.d_uBoundsU(2) = bounds.upperInputBounds.dSteeringAngleU;
            obj.d_uBoundsU(3) = bounds.upperInputBounds.dBrakesU;
            obj.d_uBoundsU(4) = bounds.upperInputBounds.dVsU;
        end

        function boundsLX = getBoundsLX(obj,x)
            boundsLX = obj.d_lBoundsX - x;
        end
        
        function boundsUX = getBoundsUX(obj,x) 
            boundsUX = obj.d_uBoundsX - x;
        end
        
        function boundsLU = getBoundsLU(obj,u)
            boundsLU = obj.d_lBoundsU - u;
        end
        
        function boundsUU = getBoundsUU(obj,u) 
            boundsUU = obj.d_uBoundsU - u;
        end
        
        function boundsLS = getBoundsLS(obj)
            boundsLS = obj.d_lBoundsS;
        end
        
        function boundsUS = getBoundsUS(obj)
            boundsUS = obj.d_uBoundsS;
        end
    end
end

